docs/auth.md

```diff
@@ -0,0 +1,176 @@
+- [Reuse authentication state](#reuse-authentication-state)
+  * [Cookies](#cookies)
+  * [Local storage](#local-storage)
+  * [Session storage](#session-storage)
+  * [Lifecycle](#lifecycle)
+  * [Example](#example)
+  * [API reference](#api-reference)
+- [Persistent authentication](#persistent-authentication)
+    - [API reference](#api-reference-1)
+- [Multi-factor authentication](#multi-factor-authentication)
+<!-- GEN:stop -->
+
+## Automate logging in
+
+The Playwright API can automate interaction with a login form. See
+[Input guide](./input.md) for more details.
+
+The following example automates login on GitHub. Once these steps are executed,
+the browser context will be authenticated.
+
+```js
+const page = await context.newPage();
+await page.goto('https://github.com/login');
+
+// Interact with login form
+await page.click('text=Login');
+await page.fill('input[name="login"]', USERNAME);
+await page.fill('input[name="password"]', PASSWORD);
+await page.click('text=Submit');
+// Verify app is logged in
+```
+
+These steps can be executed for every browser context. However, redoing login
+for every test can slow down test execution. To prevent that, we will reuse
+existing authentication state in new browser contexts.
+
+## Reuse authentication state
+
+Web apps use cookie-based or token-based authentication, where authenticated
+state is stored as [cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies)
+or in [local storage](https://developer.mozilla.org/en-US/docs/Web/API/Storage).
+The Playwright API can be used to retrieve this state from authenticated contexts
+and then load it into new contexts.
+
+Cookies and local storage state can be used across different browsers. They depend
+on your application's authentication model: some apps might require both cookies
+and local storage.
+
+The following code snippets retrieve state from an authenticated page/context and
+load them into a new context.
+
+### Cookies
+
+```js
+// Get cookies and store as an env variable
+const cookies = await context.cookies();
+process.env.COOKIES = JSON.stringify(cookies);
+
+// Set cookies in a new context
+const deserializedCookies = JSON.parse(process.env.COOKIES)
+await context.addCookies(deserializedCookies);
+```
+
+### Local storage
+Local storage ([`window.localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage))
+is specific to a particular domain.
+
+```js
+// Get local storage and store as env variable
+const localStorage = await page.evaluate(() => JSON.stringify(window.localStorage));
+process.env.LOCAL_STORAGE = localStorage;
+
+// Set local storage in a new context
+const localStorage = process.env.LOCAL_STORAGE;
+await context.addInitScript(([storage]) => {
+  if (window.location.hostname === 'example.com') {
+    const entries = JSON.parse(storage);
+    Object.keys(entries).forEach(key => {
+      window.localStorage.setItem(key, entries[key]);
+    });
+  }
+}, [localStorage]);
```

Comment:
nit: can just pass `localStorage` without an array wrap. The function would be easier as well: `storage => {...}`.